src/tuner/nccl_ofi_model.c

```diff
@@ -74,167 +66,208 @@ static const float nccl_base_lat[NCCL_NUM_ALGORITHMS][NCCL_NUM_PROTOCOLS] = {
  * across both paths, and combine the parameters. fixme: shortcomings in the
  * model forced this param high to get the right answers.
  */
-#define NCCL_OFI_TUNER_NETWORK_LATENCY            (75)
+#define NCCL_OFI_TUNER_NETWORK_LATENCY            (22.5)
 
 /*
  * With EFA, we expect some fixed cost in the device. This parameter was chosen
  * by hackishly fitting against measurements.
  */
-#define NCCL_OFI_TUNER_COMPLETION_OVERHEAD        (7.0)
-
-#define NCCL_OFI_TUNER_LL_PROTO_BW_FACTOR         (0.5)
-#define NCCL_OFI_TUNER_LL128_PROTO_BW_FACTOR      (0.9375)
-#define NCCL_OFI_TUNER_SIMPLE_PROTO_BW_FACTOR     (1.0)
+#define NCCL_OFI_TUNER_COMPLETION_OVERHEAD        (4)
 
-/* XXX: This is broadly incorrect. */
 static double all_reduce_ring_cost(struct nccl_ofi_tuner_model_dims const* dims,
 				   const int proto,
-				   const size_t nBytes)
+				   const unsigned long long chan,
+				   const unsigned long long nBytes)
 {
-	double base_latency = nccl_base_lat[NCCL_ALGO_RING][proto];
-	const double p2p_latency = nccl_nvlink_lat[NCCL_ALGO_RING][proto];
-
-	const double logranks = log(dims->num_ranks) / log(2);
-	const double logbytes = log(nBytes) / log(2);
-	const double lograt = (logbytes - logranks);
-	if (floor(lograt) < 23) {
-		/* Hack until we understand this better. */
-		return INFINITY;
-	}
-
-	const double steps_per_rotation = dims->num_ranks;
-	const double num_rotations = 2;
-	const double num_steps = (num_rotations * steps_per_rotation) - 1;
-	const double num_steps_internode = (num_rotations * dims->num_nodes);
-	const double num_steps_intranode = num_steps - num_steps_internode;
 
-	double per_proto_bandwidth_overhead;
-	double net_latency = NCCL_OFI_TUNER_NETWORK_LATENCY;
-	switch (proto) {
-	case NCCL_PROTO_LL: {
-		per_proto_bandwidth_overhead = NCCL_OFI_TUNER_LL_PROTO_BW_FACTOR;
-		break;
-	}
-	case NCCL_PROTO_LL128: {
-		per_proto_bandwidth_overhead = NCCL_OFI_TUNER_LL128_PROTO_BW_FACTOR;
-		break;
+	const double base_latency = nccl_base_lat[NCCL_ALGO_RING][proto];
+	const double p2p_latency = nccl_nvlink_lat[NCCL_ALGO_RING][proto];
+	const double ranks_per_node = (1.0*dims->num_ranks) / (1.0*dims->num_nodes);
+	const double internodeSteps = 2.0*(dims->num_nodes - 1);
+	const double intranodeSteps = ((ranks_per_node-1.0)*1.0*internodeSteps);
```

Comment:
intranode steps are `total steps - inter node steps`. so, that would be like the below:

`const double intranodeSteps = (2.0*(nRanks - 1) - internodeSteps);`

Reply:
Please review the change you're proposing: 

de86135 fixup! tuner: better ring rank/msize binning